File: podcast-gen/src/generate_audio.py
import wave
from pathlib import Path

from google.genai import types

from src.client import get_client


def generate_audio(script_path: str, output_path: str = "output/podcast.wav") -> None:
    script = Path(script_path).read_text(encoding="utf-8").strip()
    client = get_client()

    response = client.models.generate_content(
        model="gemini-2.5-pro-preview-tts",
        contents=script,
        config=types.GenerateContentConfig(
            response_modalities=["AUDIO"],
            speech_config=types.SpeechConfig(
                multi_speaker_voice_config=types.MultiSpeakerVoiceConfig(
                    speaker_voice_configs=[
                        types.SpeakerVoiceConfig(
                            speaker="MC",
                            voice_config=types.VoiceConfig(
                                prebuilt_voice_config=types.PrebuiltVoiceConfig(voice_name="Aoede")
                            ),
                        ),
                        types.SpeakerVoiceConfig(
                            speaker="ゲスト",
                            voice_config=types.VoiceConfig(
                                prebuilt_voice_config=types.PrebuiltVoiceConfig(voice_name="Puck")
                            ),
                        ),
                    ]
                )
            ),
        ),
    )

    data = response.candidates[0].content.parts[0].inline_data.data

    with wave.open(output_path, "wb") as wf:
        wf.setnchannels(1)
        wf.setsampwidth(2)
        wf.setframerate(24000)
        wf.writeframes(data)

    print(f"{output_path} を生成しました ({len(data)} bytes)")

File: ai-cast/podcast-gen/src/generate-audio.ts
import { readFileSync } from "fs";
import { FileWriter } from "wav";
import { client } from "./client.js";

export async function generateAudio(
  scriptPath: string,
  outputPath: string = "output/podcast.wav",
): Promise<void> {
  const raw = readFileSync(scriptPath, "utf-8").trim();
  const scriptMatch = raw.match(/## 台本\s*\n([\s\S]*?)(?:\n---|\n## |$)/);
  const dialogue = (scriptMatch ? scriptMatch[1].trim() : raw).replaceAll("ゲスト:", "Guest:");
  const stylePrompt = `Say the following conversation in a natural podcast style at a moderate, unhurried pace. Take your time with each sentence. MC is a warm and engaging host. Guest is a knowledgeable expert who explains things clearly and deliberately. Both speakers should sound conversational and expressive, with natural pauses between thoughts and a relaxed rhythm. Do not rush.\n\n`;
  const script = stylePrompt + dialogue;

  const response = await client.models.generateContent({
    model: "gemini-2.5-pro-preview-tts",
    httpOptions: { timeout: 600_000 },
    contents: [{ role: "user", parts: [{ text: script }] }],
    config: {
      responseModalities: ["AUDIO"],
      speechConfig: {
        multiSpeakerVoiceConfig: {
          speakerVoiceConfigs: [
            {
              speaker: "MC",
              voiceConfig: {
                prebuiltVoiceConfig: { voiceName: "Aoede" },
              },
            },
            {
              speaker: "Guest",
              voiceConfig: {
                prebuiltVoiceConfig: { voiceName: "Rasalgethi" },
              },
            },
          ],
        },
      },
    },
  });

  const data = response.candidates?.[0]?.content?.parts?.[0]?.inlineData?.data;
  if (!data) {
    throw new Error("No audio data in response");
  }

  const pcmBuffer = Buffer.from(data, "base64");

  await new Promise<void>((resolve, reject) => {
    const writer = new FileWriter(outputPath, {
      channels: 1,
      sampleRate: 24000,
      bitDepth: 16,
    });
    writer.on("error", reject);
    writer.on("finish", resolve);
    writer.write(pcmBuffer);
    writer.end();
  });

  console.log(`${outputPath} を生成しました (${pcmBuffer.length} bytes)`);
}
